src/storage/btree_load.c

```diff
@@ -4852,17 +4897,112 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
-					   unique_pk, BTREE_OP_ONLINE_INDEX_IB_INSERT, NULL);
-      /* Clear the index key. */
-      pr_clear_value (p_dbvalue);
+      index_builder_loader_task *load_task = new index_builder_loader_task (btid_int->sys_btid, &class_oids[cur_class],
+									    &cur_oid, unique_pk, load_context);
 
-      if (ret != NO_ERROR)
+      load_task->set_key (p_dbvalue);
+
+      cubthread::get_manager ()->push_task (ib_workpool, load_task);
+
+      /* Increment tasks started. */
+      tasks_started++;
+
+      /* Check for possible errors. */
+      if (load_context.m_error_code != NO_ERROR)
 	{
-	  return ret;
+	  /* Also stop all threads. */
+	  cubthread::get_manager ()->destroy_worker_pool (ib_workpool);
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, load_context.m_error_code, 0);
+	  return load_context.m_error_code;
+	}
+
+      /* Clear index key. */
+      if (p_dbvalue != &dbvalue)
+	{
+	  pr_clear_value (p_dbvalue);
+	}
+    }
+
+  /* Check if the workerpool is empty */
+  do
+    {
+      if (load_context.m_error_code != NO_ERROR)
+	{
+	  /* Also stop all threads. */
+	  cubthread::get_manager ()->destroy_worker_pool (ib_workpool);
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, load_context.m_error_code, 0);
+	  return load_context.m_error_code;
 	}
 
+      /* Wait for threads to finish. */
+      thread_sleep (10);
     }
+  while (load_context.m_tasks_executed != tasks_started);
+
+  assert (load_context.m_tasks_executed == tasks_started);
+  cubthread::get_manager ()->destroy_worker_pool (ib_workpool);
 
   return ret;
 }
+
+index_builder_loader_task::index_builder_loader_task (BTID * btid, OID * class_oid, OID * oid, int unique_pk, index_builder_loader_context & load_context):m_load_context
+  (load_context)
+{
+  BTID_COPY (&m_btid, btid);
+  COPY_OID (&m_class_oid, class_oid);
+  COPY_OID (&m_oid, oid);
+  m_unique_pk = unique_pk;
+  m_load_context.m_has_error.clear ();
+  m_load_context.m_tran_index = thread_get_thread_entry_info ()->tran_index;
+}
+
+// *INDENT-OFF*
+index_builder_loader_task::~index_builder_loader_task ()
+{
+  switch_to_global_allocator_and_call <int(DB_VALUE *), DB_VALUE * >(pr_clear_value, &m_key);
+}
+// *INDENT-ON*
+
+void
+index_builder_loader_task::set_key (DB_VALUE * key)
+{
+  switch_to_global_allocator_and_call (qdata_copy_db_value, &m_key, key);
+}
+
+void
+index_builder_loader_task::execute (cubthread::entry & thread_ref)
+{
+  int ret = NO_ERROR;
+
+  /* Check for possible errors set by the other threads. */
+  if (m_load_context.m_error_code != NO_ERROR)
+    {
+      goto cleanup;
+    }
+
+  thread_ref.tran_index = m_load_context.m_tran_index;
+
+  ret = btree_online_index_dispatcher (&thread_ref, &m_btid, &m_key, &m_class_oid, &m_oid,
+				       m_unique_pk, BTREE_OP_ONLINE_INDEX_IB_INSERT, NULL);
+
+  if (ret != NO_ERROR)
+    {
+      /* Set the error. */
+      while (m_load_context.m_has_error.test_and_set (std::memory_order_acquire))
```

Comment:
Only first error code/message should be registered. That means, only first to change m_has_error flag from false to true (test_and_set returns false) registers the error.

```c
if (m_load_context.m_has_error.test_and_set ())
  {
    // error was already set; register only the first error
  }
else
  {
    // first to set error also must register it
    m_load_context.m_error_code = ret;
    // todo - error message should also be passed.
  }
```

Also, no goto cleanup. Error or no error, m_tasks_executed should be incremented.